aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp

```diff
@@ -447,6 +449,155 @@ void qavg_pool2d_nhwc_kernel(
+    int64_t input_height,
+    int64_t input_width,
+    int64_t output_height,
+    int64_t output_width,
+    int64_t nbatch,
+    int64_t channels,
+    bool align_corners) {
+  AT_DISPATCH_QINT_TYPES(
+      input.scalar_type(), "upsample_bilinear2d_nhwc", [&]() {
+        auto* idata = static_cast<scalar_t*>(input.data_ptr());
+        auto* odata = static_cast<scalar_t*>(output.data_ptr());
+        float multiplier = input.q_scale() / output.q_scale();
+        auto minimum =
+            std::numeric_limits<typename scalar_t::underlying>::lowest();
+        auto maximum =
+            std::numeric_limits<typename scalar_t::underlying>::max();
+        const auto rheight = area_pixel_compute_scale<float>(
+            input_height, output_height, align_corners);
+        const auto rwidth = area_pixel_compute_scale<float>(
+            input_width, output_width, align_corners);
+
+        for (int64_t b = 0; b < nbatch; ++b) {
+          auto* i_p = reinterpret_cast<typename scalar_t::underlying*>(
+              idata + b * input_height * input_width * channels);
+          auto* o_p = reinterpret_cast<typename scalar_t::underlying*>(
+              odata + b * output_height * output_width * channels);
+
+          for (int64_t h2 = 0; h2 < output_height; ++h2) {
+            const auto h1r = area_pixel_compute_source_index<float>(
+                rheight, h2, align_corners, /*cubic=*/false);
+
+            const int64_t h1 = h1r;
+            const int64_t h1p = (h1 < input_height - 1) ? 1 : 0;
+            const float h1lambda = h1r - h1;
+            const float h0lambda = static_cast<float>(1.) - h1lambda;
+
+            for (int64_t w2 = 0; w2 < output_width; ++w2) {
+              const auto w1r = area_pixel_compute_source_index<float>(
+                  rwidth, w2, align_corners, /*cubic=*/false);
+              const int64_t w1 = w1r;
+              const int64_t w1p = (w1 < input_width - 1) ? 1 : 0;
+
+              const float w1lambda = w1r - w1;
+              const float w0lambda = static_cast<float>(1.) - w1lambda;
+
+              int64_t c = 0;
+              // We use float32 to do the computation
+              const auto* pos1 = i_p + (h1 * input_width + w1) * channels;
+              auto* pos2 = o_p + (h2 * output_width + w2) * channels;
+              // We have to isolate this function out because the VS does not
+              // expand the macro correctly.
+              do_bilinear_on_AVX2<scalar_t>(
+                  pos1,
+                  pos2,
+                  c,
+                  input_height,
+                  input_width,
+                  output_height,
+                  output_width,
+                  channels,
+                  output.q_zero_point(),
+                  multiplier,
+                  h0lambda,
+                  h1lambda,
+                  w0lambda,
+                  w1lambda,
+                  h1p,
+                  w1p);
+              for (; c < channels; ++c) {
```

Comment:
I don't like that `c` is taken and updated by reference in `do_bilinear_on_AVX2`, it's not readily apparent `c` is updated by reading the code here. Can you instead return the new value of `c` from `do_bilinear_on_AVX2`?